dqvm/src/global-addresses.lisp

```diff
@@ -0,0 +1,181 @@
+    :type alexandria:non-negative-fixnum
+    :initform (error-missing-initform :number-of-qubits)
+    :documentation "Number of qubits in the QVM.")
+   (number-of-processes
+    :accessor number-of-processes
+    :initarg :number-of-processes
+    :documentation "Number of parallel MPI ranks.")
+   (block-size
+    :reader block-size
+    :initarg :block-size
+    :type alexandria:non-negative-fixnum
+    :documentation "Minimum number of amplitudes per single block matrix-vector multiplication. This must be an even number and is typically chosen to be equal to two times the arity of the instruction with the maximum arity in the program to be executed.")
+   (permutation
+    :reader permutation
+    :writer update-permutation
+    :initarg :permutation
+    :type permutation
+    :documentation "Last qubit permutation evaluated, stored in a format suitable for use by APPLY-QUBIT-PERMUTATION.")
+
+   ;; The following attributes are calculated during instantiation.
+   (blocks-per-process
+    :reader blocks-per-process
+    :type alexandria:non-negative-fixnum
+    :documentation "Regular blocks per process (i.e., without including a possibly remaining block).")
+   (remainder-blocks
+    :reader remainder-blocks
+    :type alexandria:non-negative-fixnum
+    :documentation "Total number of remainder blocks.")
+   (number-of-blocks
+    :reader number-of-blocks            ; XXX remove the get- prefix
+    :type alexandria:non-negative-fixnum
+    :documentation "Total number of blocks among all address tables."))
+
+  (:default-initargs
+   :number-of-processes (mpi-comm-size)
+   :block-size *default-block-size*
+   :permutation nil)
+  (:documentation "Global address table."))
+
+(defmethod initialize-instance :after ((global-addresses global-addresses) &rest initargs)
+  (declare (ignore initargs))
+
+  (assert (not (minusp (number-of-qubits global-addresses))))
+  (assert (plusp (number-of-processes global-addresses)))
+  (assert (evenp (block-size global-addresses)))
+
+  (multiple-value-bind (bpp rem) (%blocks-per-process global-addresses)
+    (setf (slot-value global-addresses 'number-of-blocks) (%number-of-blocks global-addresses)
+          (slot-value global-addresses 'blocks-per-process) bpp
+          (slot-value global-addresses 'remainder-blocks) rem)))
+
+(defmethod print-object ((global-addresses global-addresses) stream)
+  (let ((*print-readably* nil)
+        (*print-pretty* nil))
+
+    (print-unreadable-object (global-addresses stream :type t :identity t)
+
+      (format stream "~{~A ~A~^ ~}"
+              (list (prin1-to-string :number-of-qubits) (number-of-qubits global-addresses)
+                    (prin1-to-string :number-of-processes) (number-of-processes global-addresses)
+                    (prin1-to-string :block-size) (block-size global-addresses)
+                    (prin1-to-string :permutation) (permutation global-addresses))))))
+
+(defmethod copy ((global-addresses global-addresses))
```

Comment:
I highly suggest not making such functions generic, for the reasons Pitman describes: http://www.nhplace.com/kent/PS/EQUAL.html

Prefer a special purpose `COPY-GLOBAL-ADDRESSES`